R/coerce_to_messydate.R

```diff
@@ -120,30 +203,177 @@ standardise_widths <- function(dates) {
+    x <- ifelse(stringr::str_detect(x, "^([:digit:]{2})~$|^([:digit:]{2})\\?$"),
+                paste0("00", x), x)
+    x <- ifelse(stringr::str_detect(x, "^([:digit:]{3})~$|^([:digit:]{3})\\?$"),
+                paste0("0", x), x)
+  })
+  dates <- purrr::map_chr(dates, paste, collapse = ",")
   dates
 }
+
+as_bc_dates <- function(dates) {
+  dates <- ifelse(stringr::str_count(dates, "(bc|BC|Bc|bC)") == 2,
+                  st_negative_range(dates), dates)
+  dates <- ifelse(stringr::str_count(dates, "(bc|BC|Bc|bC)") > 2,
+                  st_negative_sets(dates), dates)
+  dates <- ifelse(stringr::str_count(dates, "(bc|BC|Bc|bC)") == 1,
+                  st_negative(dates), dates)
+}
+
+as_ac_dates <- function(dates) {
+  # remove after christ letters
+  dates <- stringr::str_remove_all(dates, "(ad|AD|Ad|aD)")
+  dates <- stringr::str_trim(dates, side = "both")
+}
+
+st_negative_range <- function(dates) {
+  dates <- stringr::str_remove_all(dates, "(bc|BC|Bc|bC)")
+  dates <- gsub(" ", "", dates)
+  dates <- paste0("-", strsplit(dates, "\\.\\.")[[1]][1],
+                  "..-", strsplit(dates, "\\.\\.")[[1]][2])
+}
+
+st_negative_sets <- function(dates) {
+  dates <- stringr::str_remove_all(dates, "(bc|BC|Bc|bC)")
+  dates <- gsub(" ", "", dates)
+  dates <- unlist(strsplit(dates, "\\,"))
+  dates <- ifelse(length(dates) > 1,
+                  paste0("-", paste(dates, collapse = ", -")),
+                  paste0("-", dates))
+}
+
+st_negative <- function(dates) {
+  dates <- stringr::str_remove_all(dates, "(bc|BC|Bc|bC)")
+  dates <- stringr::str_trim(dates, side = "both")
+  dates <- paste0("-", dates)
+}
+
+ask_user <- function(dates) {
+  dates <- ifelse(stringr::str_detect(dates, "^([:digit:]{2})-([:digit:]{2})-([:digit:]{2})$") &
+                    as.numeric(gsub("-", "", stringr::str_extract(dates, "^[:digit:]{2}-"))) < 32 &
+                    as.numeric(gsub("-", "", stringr::str_extract(dates, "-[:digit:]{2}-"))) < 32 &
+                    as.numeric(gsub("-", "", stringr::str_extract(dates, "[:digit:]{2}$"))) < 32,
+                  reorder_ambiguous(dates), dates)
+  input <- utils::menu(c("Yes", "No"),
+                       title = paste0("Would you like to complete ambiguous 6 digit dates?"))
+  if (input == 1) {
+    dates <- ifelse(stringr::str_detect(dates, "^([:digit:]{2})-([:digit:]{2})-([:digit:]{2})$") &
+                      as.numeric(gsub("-", "", stringr::str_extract(dates, "^[:digit:]{2}-"))) < 23,
+                    complete_ambiguous_20(dates), dates)
+    dates <- ifelse(stringr::str_detect(dates, "^([:digit:]{2})-([:digit:]{2})-([:digit:]{2})$") &
+                      as.numeric(gsub("-", "", stringr::str_extract(dates, "^[:digit:]{2}-"))) > 22,
+                    complete_ambiguous_19(dates), dates)
+  }
+  dates
+}
+
+reorder_ambiguous <- function(d) {
+  input <- utils::menu(c("YMD (Year-Month-Day)", "DMY (Day-Month-Year)",
+                         "MDY (Month-Day-Year)"),
```

Comment:
Add fourth option where user doesn't know and it is genuinely ambiguous. We then either add an ambiguous annotation to the challenging date unit (a la ISO-...) or we create a set of the compatible dates e.g. `"{2011-10-11,2011-11-10}"

Reply:
Great idea, I have added the feature so that it returns a set of the possible dates! However, I have not yet implemented the automatic completion of dates, as above, please let me know if this is want you like instead of asking users about.